keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,168 @@
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable grpc probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+https://github.com/grpc-ecosystem/grpc-health-probe
+https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Host: pod ip
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
+      initialDelaySeconds: 5
+      periodSeconds: 10
+```
+
+This healthcheck request will be identified with the following gRPC
+`User-Agent` metadata:
+
+```
+User-Agent: kube-probe/K8S_MAJOR_VER.K8S_MINOR_VER
+```
+
+Example:
+
+```
+User-Agent: kube-probe/1.22
+```
+
+### Risks and Mitigations
+
+1. Adds more code to Kubelet and surface area to Pod.Spec. *Response*: we
+   expect that this will be generally useful given broad gRPC adoption in the
+   industry.
+
+## Design Details
+
+```go
+// core/v1/types.go
+
+type Handler struct {
+  // ...
+  TCPSocket *TCPSocketAction `json...`
+
+  // GRPC specifies an action involving a TCP port.          //+
+  // +optional                                               //+
+  GRPC *GRPCAction `json...`                                 //+
+
+  // ...
+}
+
+type GRPCAction struct {
+  // Port name or number of the port to access on the container.
+  // Number must be in the range 1 to 65535.
+  Port intstr.IntOrString `json:"port" protobuf:"bytes,2,opt,name=port"`
+
+  // Service is the name of the service to place in the gRPC HealthCheckRequest
+  // (see https://github.com/grpc/grpc/blob/master/doc/health-checking.md).
+  // +optional
+  Service *string
+}
+```
+
+### Test Plan
+
+- Unit test: Add unit tests to `pkg/kubelet/prober/...`
+- e2e: Add test case to `e2e/common/node/container_probe.go`.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implement the feature.
+- Add unit and e2e tests for the feature.
+
+#### Beta
+
+- Solicit feedback from the Alpha. Validate that API is appropriate
+  for users.
+- Ensure tests are stable and passing.
+
+Depending on skew strategy:
+
+- kubelet version skew ensures all (kubelet ver, cluster ver) support
+  the feature.
+
+#### GA
+
+- Address feedback from beta
+- Close on any remaining open issues & bugs
+
+### Upgrade / Downgrade Strategy
+
+Upgrade: N/A
+
+Downgrade: gRPC probes will not be supported in a downgrade from Alpha.
+
+### Version Skew Strategy
+
+- We may not be able to graduate this widely until all kubelet version
+  skew supports the probe type.
+
```

Comment:
This template is missing PRR, which is required.

Reply:
added

Reply:
You need to grab the questions from the latest version of the template and answer those. See https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template